functions/documents.ts

```diff
@@ -1,25 +1,465 @@
+ const velocity = doc.vote.velocity
+ const charCount = doc.body.length
+ const isDraft = score < charCount
+
+ const tagScores: Record<string, number> = {}
+ const tagVelocities: Record<string, number> = {}
+ for (const tag of doc.tags) {
+  const vote = doc.tagVotes[tag]
+  const effectiveVote = vote ?? {
+   position: 0,
+   velocity: 0,
+   timestamp: doc.created,
+  }
+  tagScores[tag] = calculateScore(effectiveVote, now)
+  tagVelocities[tag] = effectiveVote.velocity
+ }
+
+ const computed: DocumentComputed = {
+  id: doc.id,
+  title: doc.title,
+  tags: doc.tags,
+  created: doc.created,
+  modified: doc.modified,
+  rev: doc.rev,
+  score,
+  velocity,
+  charCount,
+  isDraft,
+  statusMessage: computeStatusMessage({ isDraft, velocity, score, charCount }, now),
+ }
+
+ // In list mode, expose only the top tags (net score > 1) per spec.
+ if (!includeBody) {
+  const topTags = doc.tags
+   .map((t) => {
+    const vote = doc.tagVotes[t] ?? {
+     position: 0,
+     velocity: 0,
+     timestamp: doc.created,
+    }
+    return { tag: t, score: calculateScore(vote, now) }
+   })
+   .filter((x) => x.score > 1)
+   .sort((a, b) => b.score - a.score)
+   .slice(0, 3)
+   .map((x) => x.tag)
+  computed.topTags = topTags
+ }
+ if (includeBody) {
+  computed.body = doc.body
+  computed.tagScores = tagScores
+  computed.tagVelocities = tagVelocities
+ }
+ return computed
+}
+
+function json(data: unknown, status = 200) {
+ return new Response(JSON.stringify(data), {
+  status,
+  headers: {
+   'Content-Type': 'application/json',
+  },
+ })
+}
+
+function errorText(message: string, status = 400) {
+ return new Response(message, { status })
+}
+
+async function listAllDocs(kv: any): Promise<StoredDocument[]> {
+ const { store } = await import('./lib/store.js')
+ const s = store(kv)
+ const items = await s.list(DOCUMENTS_COLLECTION)
+ return items as StoredDocument[]
+}
+
+async function getDoc(kv: any, id: string): Promise<StoredDocument | null> {
+ const { store } = await import('./lib/store.js')
+ const s = store(kv)
+ const item = await s.get(DOCUMENTS_COLLECTION, id)
+ return (item as StoredDocument) ?? null
+}
+
+async function insertDoc(kv: any, doc: StoredDocument) {
+ const { store } = await import('./lib/store.js')
+ const s = store(kv)
+ await s.insert(DOCUMENTS_COLLECTION, doc.id, doc as any)
+}
+
+async function patchDoc(kv: any, id: string, patch: Partial<StoredDocument>) {
+ const { store } = await import('./lib/store.js')
+ const s = store(kv)
+ await s.patch(DOCUMENTS_COLLECTION, id, patch as any)
+}
+
+async function deleteDoc(kv: any, id: string) {
+ const { store } = await import('./lib/store.js')
+ const s = store(kv)
+ await s.delete(DOCUMENTS_COLLECTION, id)
+}
+
+function shouldHideExpiredDraft(doc: StoredDocument, now = Date.now()) {
+ const score = calculateScore(doc.vote, now)
+ const isDraft = score < doc.body.length
+ if (!isDraft) return false
+ if (doc.vote.velocity >= 0) return false
```

Comment:
We should hide drafts with score <= 0 even if velocity is >= 0, this needs to be corrected